addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/handler/MiHomeGatewayHandler.java

```diff
@@ -0,0 +1,575 @@
+        setProperty(dataObj, DeviceConstants.GATEWAY_FIRMWARE_VERSION_KEY, props,
+                MiHomeBindingConstants.PROPERTY_FIRMWARE_VERSION);
+        setProperty(dataObj, DeviceConstants.DEVICE_ID_KEY, props, MiHomeBindingConstants.PROPERTY_DEVICE_ID);
+        setProperty(dataObj, DeviceConstants.USER_ID_KEY, props, MiHomeBindingConstants.PROPERTY_USER_ID);
+        setProperty(dataObj, DeviceConstants.GATEWAY_MAC_ADDRESS_KEY, props,
+                MiHomeBindingConstants.PROPERTY_MAC_ADDRESS);
+        updateProperties(props);
+    }
+
+    /**
+     * Gets the property values from the device information one by one and puts them in the map
+     *
+     * @param dataObj - JSON data object which the properties should be get from
+     * @param devicePropertyToGet - the specific property to get from the data
+     * @param mapToPut - a map where the taken property will be put
+     * @param thingPropertyToSet - name of the property that is put in the map
+     */
+    private void setProperty(JsonObject dataObj, String devicePropertyToGet, Map<String, String> mapToPut,
+            String thingPropertyToSet) {
+        JsonElement property = dataObj.get(devicePropertyToGet);
+        if (!property.isJsonNull()) {
+            mapToPut.put(thingPropertyToSet, property.getAsString());
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (channelUID.getId().equals(MiHomeBindingConstants.CHANNEL_LAST_SEEN)) {
+            if (command instanceof DateTimeType) {
+                State lastSeenState = DateTimeType.valueOf(command.toString());
+                updateState(channelUID, lastSeenState);
+            }
+        }
+    }
+
+    @Override
+    public void thingUpdated(Thing thing) {
+        /*
+         * The default behavior is not wanted here because disposing and initializing the
+         * handler would result in a new device id.
+         */
+        String newLabel = thing.getLabel();
+        if (!gatewayLabel.equals(newLabel)) {
+            this.thing.setLabel(newLabel);
+            logger.warn(CHANGING_LABEL_MESSAGE);
+        }
+    }
+
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
+        logger.debug("Editing the thing configuration...");
+        Configuration config = editConfiguration();
+        username = (String) configurationParameters.get(MiHomeBindingConstants.CONFIG_USERNAME);
+        config.put(MiHomeBindingConstants.CONFIG_USERNAME, username);
+        gatewayCode = (String) configurationParameters.get(MiHomeBindingConstants.CONFIG_GATEWAY_CODE);
+        config.put(MiHomeBindingConstants.CONFIG_GATEWAY_CODE, gatewayCode);
+        password = configurationParameters.get(MiHomeBindingConstants.CONFIG_PASSWORD).toString();
+        config.put(MiHomeBindingConstants.CONFIG_PASSWORD, password);
+        updateInterval = Long
```

Comment:
You are now actually doing the following:
- converting a BigDecimal to a string by calling toString on it 
- parse the string to a Long-object by passing it to the Long.valueOf
- converting the Long object by auto-unboxing it by assigning to a long field